src/main/java/com/typesafe/netty/http/pipelining/HttpPipeliningHandler.java

```diff
@@ -1,110 +1,221 @@
+            ctx.fireChannelReadComplete();
+        }
     }
 
     @Override
-    public void messageReceived(final ChannelHandlerContext ctx, final MessageEvent e) {
-        final Object msg = e.getMessage();
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
         if (msg instanceof HttpRequest) {
-            ctx.sendUpstream(new OrderedUpstreamMessageEvent(sequence++, e.getChannel(), msg, e.getRemoteAddress()));
+            receiveSequence++;
+
+            if (inFlight.incrementAndGet() > inFlightRequestsHighWatermark) {
+                highWatermarkExceeded = true;
+            }
+            HttpRequest request = (HttpRequest) msg;
+            ctx.fireChannelRead(new SequencedHttpRequest(receiveSequence, request));
+
         } else {
-            ctx.sendUpstream(e);
+            ctx.fireChannelRead(msg);
         }
     }
 
     @Override
-    public void handleDownstream(ChannelHandlerContext ctx, ChannelEvent e)
-            throws Exception {
-        if (e instanceof OrderedDownstreamChannelEvent) {
-
-            boolean channelShouldClose = false;
-
-            synchronized (holdingQueue) {
-                if (holdingQueue.size() < maxEventsHeld) {
-
-                    final OrderedDownstreamChannelEvent currentEvent = (OrderedDownstreamChannelEvent) e;
-                    holdingQueue.add(currentEvent);
-
-                    while (!holdingQueue.isEmpty()) {
-                        final OrderedDownstreamChannelEvent nextEvent = holdingQueue.peek();
-                        if (nextEvent.getOrderedUpstreamMessageEvent().getSequence() != nextRequiredSequence |
-                                nextEvent.getSubsequence() != nextRequiredSubsequence) {
-                            break;
-                        }
-                        holdingQueue.remove();
-                        ctx.sendDownstream(nextEvent.getChannelEvent());
-                        if (nextEvent.isLast()) {
-                            ++nextRequiredSequence;
-                            nextRequiredSubsequence = 0;
-                        } else {
-                            ++nextRequiredSubsequence;
-                        }
-                    }
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        if (msg instanceof SequencedOutboundMessage) {
+            SequencedOutboundMessage sequenced = (SequencedOutboundMessage) msg;
+
+            synchronizedWrite(ctx, sequenced, promise);
+        } else {
+            ctx.write(msg, promise);
+        }
+    }
+
+    private void progressToNextSendingSequence(ChannelHandlerContext ctx) {
+        currentlySendingSequence++;
+
+        int inFlight = this.inFlight.decrementAndGet();
+        // If we're now at the low water mark, issue a read for the next request
+        if (highWatermarkExceeded && inFlight == inFlightRequestsLowWatermark) {
+            ctx.read();
```

Comment:
another problem I think is that ``ctx.read()`` just issue a read op.not a complete read of HttpRequest,and if there is a read op is pending,this read will be ignored .so I think we should trigger the next read() in the ```channelReadComplete```.

the ```channelReadComplete``` is just a callback of the ```read``` op complete but not a whole HttpRequest message.So that's it.